### ISSUE TITLE
Nessie client randomly fails when connecting to TLS ingress

### ISSUE THREAD

**Issue by tmnd1991** [projectnessie/nessie#4399]
**What happened**:

When issuing commands through the Nessie cli, oftentimes the client errors out as follows:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/requests/adapters.py", line 440, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 785, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.9/dist-packages/urllib3/util/retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.9/dist-packages/urllib3/packages/six.py", line 769, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/local/lib/python3.9/dist-packages/urllib3/connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/nessie", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/pynessie/decorators.py", line 62, in wrapper
    f(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/pynessie/commands/remote.py", line 40, in show
    click.echo("Default branch: " + ctx.nessie.get_reference(None).name)
  File "/usr/local/lib/python3.9/dist-packages/pynessie/client/nessie_client.py", line 107, in get_reference
    else get_default_branch(self._base_url, self._auth, self._ssl_verify)
  File "/usr/local/lib/python3.9/dist-packages/pynessie/client/_endpoints.py", line 138, in get_default_branch
    return cast(dict, _get(base_url + "/trees/tree", auth, ssl_verify=ssl_verify))
  File "/usr/local/lib/python3.9/dist-packages/pynessie/client/_endpoints.py", line 41, in _get
    r = requests.get(url, headers=_get_headers(), verify=ssl_verify, params=params, auth=auth)
  File "/usr/local/lib/python3.9/dist-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/requests/sessions.py", line 529, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.9/dist-packages/requests/sessions.py", line 645, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/requests/adapters.py", line 501, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

**What you expected to happen**:

The client to work 100% of the time.

**How to reproduce it (as minimally and precisely as possible)**:

**Details**:
- Nessie server type (docker/built from source) and version: Nessie 0.30.0, docker image on EKS
- Client type (Ex: UI/Spark/pynessie ...) and version: Nessie 0.30.0 both from Spark and pynessie, no errors from the UI.

Since Nessie server is deployed on EKS we have an AWS NLB (with TLS enabled) and nginx acting as an ingress. The ingress configuration is the same as other http server we expose and so far we never had any problem. The Nessie UI works flawlessly and whenever we encountered this problem we never saw any log on Nessie server.
Nessie server is deployed without enabling TLS, TLS is enabled only at ingress side.

Pointing directly to the internal service name (without TLS nor ingress) works well, but only for applications that run on the same EKS cluster, which a big limitation

**Comment by snazy** [comment]
As far as I could figure out, there were a few issues in urllib(3) and related Python code regarding a missing ALPN extension for TLS. However, that would probably mean, that the requests do always fail, which doesn't seem to be what happens in your case.

EKS however may have an issue in the network stack with SNAT (and DNAT) (see [here](https://stackoverflow.com/a/62777751)). The mentioned thread and the linked posts are quite old, but maybe it helps.

As TLS seems to be terminated at a load-balancer, the Nessie server's only dealing with "plain HTTP". Since you mentioned that Nessie works fine when accessed directly, I doubt that the issue's related to the Nessie server.

I think it's related to the combination of Python, urllib(3) and EKS + NLB terminating TLS. Can you share more information like the actual Python runtime version, about the actual EKS setup and such?

**Comment by tmnd1991** [comment]
Sure, it's very likely that the problem is on the Nessie client side and the python library/runtime used to perform http requests, because other clients (like the browser) work fine with that setup.

Python runtime is:

```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
```

urllib version:

```
$ pip show urllib3
Name: urllib3
Version: 1.26.9
```

requests version:

```
$ pip show requests
Name: requests
Version: 2.27.1
```

**Comment by snazy** [comment]
Could you try with the most recent Python 3.9.13?

> very likely that the problem is on the Nessie client side and the python library/runtime

Could be, but could also be something related to TLS and "all the things" involved w/ TLS on the way.
One way to narrow it down could be to run `curl` in a loop against the `/api/v1/trees/tree` REST endpoint from the same machine that runs the Python Nessie CLI. I suspect that `curl` works fine, but better safe than sorry.
And maybe a similar test with a Python script using the same urllib3 version.

**Comment by snazy** [comment]
Oh - and check which OpenSSL version's installed (should be 1.1.1).

**Comment by tmnd1991** [comment]
Hi @snazy , I kept digging into that and I found out that from the host I was experiencing the error, I also experienced failures using curl, only when connecting to the ingress, while I didn't connecting to the internal service name, while I wasn't experiencing any issue from my laptop (from a container with the same image).

I think we can close this now, since it is probably a problem in our network configuration. Thank you very much for your help.
